[Inductor][NV Universal GEMM]  Fix autotuning for GEMM inputs that are views of the same buffer

NikhilAPatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171556
* #171360
* #171205
* __->__ #171363
* #171362
* #170623

When GEMM inputs are views of the same underlying buffer (common in fused QKV attention patterns where Q, K, V are slices of a single fused tensor), compile-time autotuning ends up building only one example tensor:
  - The NV Universal GEMM wrapper hands `generate_kernel_call()` only argument names. For `ReinterpretView` nodes, `get_name()` is the base buffer name, not a unique view identifier.
  - If both A and B matrices are views of the same tensor (e.g., `fused_qkv`), they share that name.
  - When the autotuner later fabricates example tensors, it sees duplicate names and generates a single tensor instead of
    two distinct views.
  - Benchmarks then run with the wrong shapes/strides, causing failures.


By passing the `ReinterpretView` itself as r`aw_arg` (instead of just the underlying buffer name), `autotune_at_compile_time` can:
  - Access each view’s layout information (size, stride, offset).
  - Generate separate example tensors per view, even when they share storage.
  - Ensure every tensor used during benchmarking reflects the correct view layout.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

NikhilAPatel committed:
Pass the ReinterpretView as raw_arg so autotune_at_compile_time can use it to generate example tensors

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor][NV Universal GEMM]  Fix autotuning for GEMM inputs that are views of the same buffer"


When GEMM inputs are views of the same underlying buffer (common in fused QKV attention patterns where Q, K, V are slices of a single fused tensor), the autotuning framework has a deduplication bug:

1. AlgorithmSelectorCache.get_inputs() deduplicates inputs by calling get_name() on each input node
2. For ReinterpretView nodes, get_name() returns the underlying buffer name, not a unique view identifier
3. If both A and B matrices are views of the same buffer (e.g., fused_qkv), they have the same name
4. The deduplication results in only 1 tensor being passed when 2 are expected
5. This causes benchmark failures or incorrect results


Looking at the code in nv_universal_gemm_kernel.py:
```
  for _, input_node in self._template_input_args:
      reinterpret_view = self._get_reinterpret_view(input_node)
      if reinterpret_view is not None:
          call_args.append(reinterpret_view.codegen_reference())
          # Pass the ReinterpretView as raw_arg so autotune_at_compile_time
          # can use it to generate example tensors
          raw_args.append(reinterpret_view)
      else:
          call_args.append(input_node.get_name())
          raw_args.append(input_node)
```

By passing the ReinterpretView itself as raw_arg (instead of the underlying buffer), autotune_at_compile_time can:
1. Access the view's layout information (size, stride, offset)
2. Generate separate example tensors for each view, even if they share the same underlying buffer name
3. Each tensor gets the correct shape/stride/offset from the ReinterpretView's layout

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor][NV Universal GEMM]  Fix autotuning for GEMM inputs that are views of the same buffer"


When GEMM inputs are views of the same underlying buffer (common in fused QKV attention patterns where Q, K, V are slices of a single fused tensor), the autotuning framework has a deduplication bug:

1. AlgorithmSelectorCache.get_inputs() deduplicates inputs by calling get_name() on each input node
2. For ReinterpretView nodes, get_name() returns the underlying buffer name, not a unique view identifier
3. If both A and B matrices are views of the same buffer (e.g., fused_qkv), they have the same name
4. The deduplication results in only 1 tensor being passed when 2 are expected
5. This causes benchmark failures or incorrect results


Looking at the code in nv_universal_gemm_kernel.py:
```
  for _, input_node in self._template_input_args:
      reinterpret_view = self._get_reinterpret_view(input_node)
      if reinterpret_view is not None:
          call_args.append(reinterpret_view.codegen_reference())
          # Pass the ReinterpretView as raw_arg so autotune_at_compile_time
          # can use it to generate example tensors
          raw_args.append(reinterpret_view)
      else:
          call_args.append(input_node.get_name())
          raw_args.append(input_node)
```

By passing the ReinterpretView itself as raw_arg (instead of the underlying buffer), autotune_at_compile_time can:
1. Access the view's layout information (size, stride, offset)
2. Generate separate example tensors for each view, even if they share the same underlying buffer name
3. Each tensor gets the correct shape/stride/offset from the ReinterpretView's layout

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor][NV Universal GEMM]  Fix autotuning for GEMM inputs that are views of the same buffer"


When GEMM inputs are views of the same underlying buffer (common in fused QKV attention patterns where Q, K, V are slices of a single fused tensor), the autotuning framework has a deduplication bug:

1. AlgorithmSelectorCache.get_inputs() deduplicates inputs by calling get_name() on each input node
2. For ReinterpretView nodes, get_name() returns the underlying buffer name, not a unique view identifier
3. If both A and B matrices are views of the same buffer (e.g., fused_qkv), they have the same name
4. The deduplication results in only 1 tensor being passed when 2 are expected
5. This causes benchmark failures or incorrect results


Looking at the code in nv_universal_gemm_kernel.py:
```
  for _, input_node in self._template_input_args:
      reinterpret_view = self._get_reinterpret_view(input_node)
      if reinterpret_view is not None:
          call_args.append(reinterpret_view.codegen_reference())
          # Pass the ReinterpretView as raw_arg so autotune_at_compile_time
          # can use it to generate example tensors
          raw_args.append(reinterpret_view)
      else:
          call_args.append(input_node.get_name())
          raw_args.append(input_node)
```

By passing the ReinterpretView itself as raw_arg (instead of the underlying buffer), autotune_at_compile_time can:
1. Access the view's layout information (size, stride, offset)
2. Generate separate example tensors for each view, even if they share the same underlying buffer name
3. Each tensor gets the correct shape/stride/offset from the ReinterpretView's layout

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

drisspg commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -3010,6 +3010,9 @@ def infer_arg_by_inputs(raw_keys, raw_args, idx, reused_args):
                 assert len(raw_args) == len(call_args), (
                     "call_args and raw_args do not match"
                 )
+                # If raw_args is provided but raw_keys is not, fill with None
```
this comment isn't helpful can you explain why?

drisspg replied:
or form speakign offline, can we create dummy raw_keys in the lowering

NikhilAPatel committed:
Update on "[Inductor][NV Universal GEMM]  Fix autotuning for GEMM inputs that are views of the same buffer"


When GEMM inputs are views of the same underlying buffer (common in fused QKV attention patterns where Q, K, V are slices of a single fused tensor), compile-time autotuning ends up building only one example tensor:
  - The NV Universal GEMM wrapper hands `generate_kernel_call()` only argument names. For `ReinterpretView` nodes, `get_name()` is the base buffer name, not a unique view identifier.
  - If both A and B matrices are views of the same tensor (e.g., `fused_qkv`), they share that name.
  - When the autotuner later fabricates example tensors, it sees duplicate names and generates a single tensor instead of
    two distinct views.
  - Benchmarks then run with the wrong shapes/strides, causing failures.


By passing the `ReinterpretView` itself as r`aw_arg` (instead of just the underlying buffer name), `autotune_at_compile_time` can:
  - Access each view’s layout information (size, stride, offset).
  - Generate separate example tensors per view, even when they share storage.
  - Ensure every tensor used during benchmarking reflects the correct view layout.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

drisspg reviewed (approved):
looks great

NikhilAPatel commented:
@pytorchbot merge -f "Unrelated failures"
